### DISCUSSION TITLE
Получение токена с помощью логина и пароля

### DISCUSSION THREAD

**Discussion by MarshalX** [MarshalX/yandex-music-api#513] [Покажите и расскажите]
## ‼️ Вся информация о том как получить токен теперь тут: https://ym.marshal.dev/token

**Comment by glebliutsko** [comment]
# Самый простой через браузер
1. *(Опционально)* Открываем DevTools в браузере и на вкладке Network включаем троттлинг.
2. Переходим по ссылке https://oauth.yandex.ru/authorize?response_type=token&client_id=23cabbbdc6cd418abb4b39c32c41195d
3. Авторизуемся при необходимости и предоставляем доступ
4. Браузер перенаправит на адрес вида `https://music.yandex.ru/#access_token=AQAAAAAYc***&token_type=bearer&expires_in=31535645`. Очень быстро произойдет редирект на другую страницу, поэтому нужно успеть скопировать ссылку.
   ![изображение](https://user-images.githubusercontent.com/40576167/167846948-c288a9f1-b671-4edc-a851-9f50f339f5d8.png)
6. Ваш токен, то что находится после `access_token`.

## Как включить троттлинг
- [Chrome](https://www.browserstack.com/guide/how-to-perform-network-throttling-in-chrome)
- [Firefox](https://firefox-source-docs.mozilla.org/devtools-user/network_monitor/throttling/index.html)

**Reply by pegakmop** [reply]
1) способ:
Вставлю свои 5 копеек для получения токена на платформе iOS есть замечательное приложение shortcuts(быстрые команды) от apple, чтобы не пытаться быть ковбоем в шляпе и проверять быстроту и скорость реакции на редирект, создал команду которая получает токен в пару действий при этом никуда торопиться не нужно, не спеша получая токен, скачать команду можно по ссылке https://www.icloud.com/shortcuts/a712ce8beb1744bdba51330ed42b2e43 приятного получения токена  юзерам 👌 устарело…

2) новый способ:🆕13.06.2025
Считаю команды устаревшими, нашел новый способ получения токена, откройте страницу и увидите инструкцию: https://pegakmop.github.io/yatoken

**Reply by TOMMy-Net** [reply]
На счет IOS команды, это наверное самый быстрый и простой способ!!!

**Reply by nubovik01** [reply]
@pegakmop +rep

**Comment by NicWeiss** [comment]
Автоматизированный вариант получения через selenium
Поднимает инстанцию Chrome, в которой вам необходимо авторизоваться на в паспорте яндекса. Далее из XHR вычленяется токен авторизации

```
import json
from time import sleep

from selenium import webdriver
from selenium.webdriver import DesiredCapabilities
from selenium.webdriver.remote.command import Command
from webdriver_manager.chrome import ChromeDriverManager


def is_active(driver):
    try:
        driver.execute(Command.GET_ALL_COOKIES)
        return True
    except Exception:
        return False


def get_token():
    # make chrome log requests
    capabilities = DesiredCapabilities.CHROME
    capabilities["loggingPrefs"] = {"performance": "ALL"}
    capabilities['goog:loggingPrefs'] = {'performance': 'ALL'}
    driver = webdriver.Chrome(desired_capabilities=capabilities,
                              executable_path=ChromeDriverManager().install())
    driver.get(
        "https://oauth.yandex.ru/authorize?response_type=token&client_id=23cabbbdc6cd418abb4b39c32c41195d")

    token = None

    while token == None and is_active(driver):
        sleep(1)
        try:
            logs_raw = driver.get_log("performance")
        except:
            pass

        for lr in logs_raw:
            log = json.loads(lr["message"])["message"]
            url_fragment = log.get('params', {}).get('frame', {}).get('urlFragment')

            if url_fragment:
                token = url_fragment.split('&')[0].split('=')[1]

    try:
        driver.close()
    except:
        pass

    return token


print(get_token())

```

**Reply by xoxloviwan** [reply]
```bash
#!/bin/bash
/opt/google/chrome/chrome --enable-logging=stderr 'https://oauth.yandex.ru/authorize?response_type=token&client_id=23cabbbdc6cd418abb4b39c32c41195d' -v=4 > tmp.txt 2>&1 &
sleep 2
grep -e 'access_token=' tmp.txt
```
Если лень качать драйвер/селениум (как мне), можно так (я слоупок и тротлинг мне не помог)
только после этого нужно загасить хром иначе много наспамит в tmp.txt - за пару минут написал 1,5 Гб

**Reply by Stranix81** [reply]
не подскажешь, какую версию webdriver_manager надо юзать? установил через pip последнюю - выдаёт ошибку в "driver = webdriver.chrome..."
ошибка: WebDriver.__init__() got an unexpected keyword argument 'desired_capabilities'
так понимаю, новая версия webdriver_manager не поддерживает такие аргументы.

**Reply by andrey-zhilin** [reply]
@Stranix81 мне помогло
`install selenium==4.9.0`

**Reply by Stranix81** [reply]
@andrey-zhilin спасибо

**Reply by SBRDIGITAL** [reply]
> Автоматизированный вариант получения через selenium Поднимает инстанцию Chrome, в которой вам необходимо авторизоваться на в паспорте яндекса. Далее из XHR вычленяется токен авторизации
> 
> ```
> import json
> from time import sleep
> 
> from selenium import webdriver
> from selenium.webdriver import DesiredCapabilities
> from selenium.webdriver.remote.command import Command
> from webdriver_manager.chrome import ChromeDriverManager
> 
> 
> def is_active(driver):
>     try:
>         driver.execute(Command.GET_ALL_COOKIES)
>         return True
>     except Exception:
>         return False
> 
> 
> def get_token():
>     # make chrome log requests
>     capabilities = DesiredCapabilities.CHROME
>     capabilities["loggingPrefs"] = {"performance": "ALL"}
>     capabilities['goog:loggingPrefs'] = {'performance': 'ALL'}
>     driver = webdriver.Chrome(desired_capabilities=capabilities,
>                               executable_path=ChromeDriverManager().install())
>     driver.get(
>         "https://oauth.yandex.ru/authorize?response_type=token&client_id=23cabbbdc6cd418abb4b39c32c41195d")
> 
>     token = None
> 
>     while token == None and is_active(driver):
>         sleep(1)
>         try:
>             logs_raw = driver.get_log("performance")
>         except:
>             pass
> 
>         for lr in logs_raw:
>             log = json.loads(lr["message"])["message"]
>             url_fragment = log.get('params', {}).get('frame', {}).get('urlFragment')
> 
>             if url_fragment:
>                 token = url_fragment.split('&')[0].split('=')[1]
> 
>     try:
>         driver.close()
>     except:
>         pass
> 
>     return token
> 
> 
> print(get_token())
> ```

Приветствую! Ваш вариант устаревший и уже не работает! Вот актуальная версия:
https://github.com/SBRDIGITAL/get_yandex_music_token_with_selenium.git

**Comment by suntrackspb** [comment]
Мне тоже не помогал тротлинг, просто не грузил страницу...  поэтому добавлю свои 5 копеек... 

1. Открываете DevTools (инструменты разработчика) в любом браузере.
2. На вкладке **Network** (Сеть): в хроме и ему подобным: находим и ставим галочку **Preserve log,** в файрфокс жмем шестеренку и там **неприрывный лог**
3. Жмык: https://oauth.yandex.ru/authorize?response_type=token&client_id=23cabbbdc6cd418abb4b39c32c41195d
4. В хроме видим поле Filter, В firefox оно называется поиск url, пишем в это поле "**auth?**" (БЕЗ КАВЫЧЕК)
5.  Найдется запрос типо такого: _**auth?external-domain=music.yandex.ru&overembed=no&__t=1678486679211**_, тыкаем в него.
6. Откроется новый фрейм там на владке **Headers**, в самом низу, в параметре **X-Retpath-Y**: и будет ваша ссылка с ключом.

ВАЖНО: не забудьте когда копируете ключ что он начинается после символов "access_token%3D" так как это кодировка знака  =

**Comment by savely-krasovsky** [comment]
@MarshalX ничего не понимаю. Зачем в итоге такие страшные схемы с мобильным приложением (!) и расширением для браузера (!), если это обычный OAuth 2.0? Я на коленке за полчаса написал элементарный PoC на базе [@OpenIDAppAuth-js](https://github.com/openid/AppAuth-JS), который полностью в браузере получает access_token.

Объясните, мб я чего-то не понимаю?

**Reply by MarshalX** [reply]
Обсудили это с тобой в телеграм чате. Продублирую тут кратко. Нет возможности создать свое oauth приложение с правами к Яндекс Музыке. Следовательно нет возможности настроить redirect_url (callback uri) на свой собственный сервер. Приходится перехватывать запросы на redirect_url яндекса

**Comment by andronix1** [comment]
Метод из приложения
https://github.com/MarshalX/yandex-music-api/issues/601

**Comment by Snoxc** [comment]


**Reply by andronix1** [reply]
То есть не работает в библиотеке для музыки?

**Reply by Elena1294** [reply]
У меня тоже токен не работает, ошибка 401, в чем может быть проблема?

**Reply by andronix1** [reply]
Можно посмотреть реализацию в репозитории https://github.com/MarshalX/yandex-music-token в ветке v2.

**Reply by Elena1294** [reply]
ветка недоступна 404 - page not found

**Reply by MarshalX** [reply]
Не советую v2, давно помер. Расширения и андроид приложение лучше всего

**Reply by andronix1** [reply]
Ой. Мне показалось, что v2 это из обсуждения #601. После получения первого токена получается токен для Я.Музыки через запрос на https://mobileproxy.passport.yandex.net/1/token?app_platform=android?

**Reply by MarshalX** [reply]
Возможно оно похоже

**Comment by wokalek** [comment]
А если использовать обычную авторизацию OAuth яндекса путём создания у них приложения? При таком подходе не будет просто прав на работу с музыкой?

**Comment by egoarka** [comment]
Так тоже можно, запускать в консоли по адресу https://oauth.yandex.ru/

```javascript
fetch(
  "https://oauth.yandex.ru/authorize?response_type=token&client_id=23cabbbdc6cd418abb4b39c32c41195d",
)
  .then((a) => a.text())
  .then((a) => console.log(a.match(/access_token=(.*?)&/)[1]));
```

**Comment by klauday** [comment]
Как получить токен для музыки сейчас ?

**Reply by MarshalX** [reply]
Аналогично вчерашнему способу

**Reply by klauday** [reply]
> Аналогично вчерашнему способу 

А что за прикол получил токен и делаю любой запрос и там пишет unavailable for legal reasons

**Reply by MarshalX** [reply]
Нужно делать запросы со стран где официалаьно есть ЯМ. Иначе купить подписку

**Comment by alahmatiy** [comment]
А это... приложения же можно создавать теперь на https://oauth.yandex.ru

**Reply by nickolai** [reply]
> А это... приложения же можно создавать теперь на https://oauth.yandex.ru

Спасибо за подсказку. Что потребуется: client_id?

**Reply by alahmatiy** [reply]
Да. Авторизация со своим client_id приложения.

```python
def get_auth_params():
    auth_url = ("https://oauth.yandex.ru/authorize?response_type=token&client_id={app_id}".format(app_id=APP_ID))
    webbrowser.open_new_tab(auth_url)
    redirected_url = input("Paste here url you were redirected:\n")
    aup = parse_qs(redirected_url)
    aup['access_token'] = aup.pop('https://test.ru/#access_token')
    return aup['access_token'][0], aup['cid'][0]
```

**Reply by MarshalX** [reply]
@alahmatiy создавать приложения можно было всегда, а вот запрасить права для ЯМ было невозможно. Без этих прав своё приложение бесполезно. Если права для ЯМ стало возможно запросить, нужны скриншоты, как минимум

**Reply by alahmatiy** [reply]
А там и нет прав на ЯМ по моему в перечне. Но с полученным токеном я выполняю запросы АПИ и они работают. Может я чего-то не совсем понял конечно, но мой скрипт работает и все ок вроде.

**Reply by andreyvolokitin** [reply]
@alahmatiy треки можно прослушивать полностью с таким токеном? Некоторые запросы АПИ не требуют токена в принципе, потому они могут работать независимо от токена. Работают ли запросы, которые требуют токен?

**Comment by intexcor** [comment]
https://oauth.yandex.ru/authorize?response_type=token&client_id=1a6990aa636648e9b2ef855fa7bec2fb

самый простой способ

**Reply by krax1337** [reply]
Спасибо! Прилинкую, то что это способ из Я.Облака https://yandex.cloud/ru/docs/cli/quickstart#initialize, относительно безопасно, но токен лучше никому не сливать

**Reply by bonlime** [reply]
Тот факт что токены из я.облака работают значит что это по сути самый простой способ всегда иметь свежий без необходимости делать эмуляцию браузера. можно создать service account на YC и через него получать токен в любой нужные момент

**Comment by intexcor** [comment]
А какие запросы не требуют токена? Запросы инфы про трек?


чт, 1 мая 2025 г. в 19:58, andreyvolokitin ***@***.***>:

**Reply by andreyvolokitin** [reply]
Инфа про трек не требует токена. Точно требует токена ссылка на полный трек. Мне кажется токена требуют эндпоинты, для которых указан `OAuth` тут, например: https://github.com/acherkashin/yandex-music-open-api/blob/main/src/yandex-music.yaml#L221

**Comment by pegakmop** [comment]
> ### ‼️ Как получить токен
> **Своё OAuth приложение создать нельзя.** Единственный вариант это использовать приложения официальных клиентов Яндекс.Музыка.

Смотрите какой легкий способ на iOS нашел: https://pegakmop.github.io/yatoken возможно на пк тоже сработает с тампермонкей установив скрипт, на счет андроида не знаю может кто-то проверит)))

**Comment by Yellastro2** [comment]
Авторизация по андроид Яндекс SDK больше не работает по схеме из приложения github.com/MarshalX/yandex-music-token/tree/main/android-app - видимо аннулировали ключ того приложения. Как достали тот ключ кто-нибудь знает? Где искать новый? В манифесте последней версии официального приложения никаких ключей не вижу

**Comment by MarkZaytsev** [comment]
Возможно кому-то будет полезно:

Я получил токен через своё приложение созданное на https://oauth.yandex.ru, это довольно просто и занимает пару минут. Но с этим токеном прога скачивает все трэки в качестве `preview` (128 kbit). Именно такой формат возвращает запрос `track_info`, не зависимо какое качество было указано в запросе. Аналогичная проблема с токеном из Я.Облака (https://yandex.cloud/ru/docs/cli/quickstart#initialize).

Получать токены через сторонние apk и телеграм боты с вводом логин/пароля мне стрёмно совсем.

В итоге я набрёл на простое и эффективное решение от @Sorrow446: https://github.com/Sorrow446/Yandex-Music-Downloader/tree/token_extractor
Эта штука тупо достаёт токен из кэша десктопного плеера, а код помещается в пару фунок которые я действительно легко мог понять и проверить. Ещё у меня сложилось подозрение, что ограничение на `preview` качество распространяется именно на токены из веб версии судя по документации в её репозитории, но я уже не стал с этим разбираться.

Я использовал его на Fedora Linux с десктопной версией https://github.com/cucumber-sp/yandex-music-linux.
Токен получил моментально без танцев с бубном, с ним скачиваются flac/m4a и всё остальное в высоком качестве.

**Comment by Stmol** [comment]
Написал небольшой веб GUI по мотивам питон скрипта для извлечения токена, можно развернуть локально или заселфхостить для друзей: https://github.com/Stmol/yandex-oauth-token
